File: lib/toastbot.js
var fs = require('fs');
var http = require('http');
var irc = require('irc');
var path = require('path');
var spawn = require('child_process').spawn;


var __author__ = 'Daniel Lindsley <daniel@toastdriven.com>';
var __version__ = [0, 1, 0];


var zero_pad = function(value, width) {
  var adjusted = value.toString();
  var width = width || 2;
  
  while(adjusted.length < width) {
    adjusted = '0'+adjusted;
  }
  
  return adjusted;
};


var method = function(fn) {
  return function() {
    return fn.apply(this, [this].concat([].slice.call(arguments)));
  }
};

var handler = function(fn) {
  var retval = function() {
    return fn.apply(this, arguments);
  };
  var enabled = true;

  retval.enabled = function() {
    return enabled;
  };

  retval.disable = function() {
    enabled = false;
  };

  retval.enable = function() {
    enabled = true;
  };

  return retval;
};

var Toastbot = method(function(self, nick, channel, options) {
  options = options || {};
  
  self.nick = nick;
  self.channel = channel;
  self.client = null;
  
  self.server = options.server || 'irc.freenode.net';
  self.username = options.username || self.nick;
  self.realname = options.realname || 'ToastBot';
  self.debug = options.debug || false;
  self.log_dir = options.log_dir || path.join(path.dirname(path.dirname(__filename)), 'logs');
  self.variants = options.variants || [
    self.nick+': ',
    self.nick+', ',
    self.nick+'- ',
    self.nick+' - '
  ];
});

// ==============================================
// Initialization and bookkeeping methods
// ==============================================

Toastbot.prototype.setup = method(function(self) {
  self.ensure_log_directory();
  
  self.client = new irc.Client(self.server, self.nick, {
    userName: self.username,
    realName: self.realname,
    debug: self.debug,
    channels: [self.channel]
  });
  
  self.client.on('join'+self.channel, function() {
    self.log('Joined '+self.channel);
  });
  self.client.on('message'+self.channel, function(nick, text) {
    self.handle_message(nick, text);
  });
  self.client.on('pm', function(nick, text) {
    self.handle_pm(nick, text);
  });
  self.client.on('error', function(raw_message) {
    console.log(raw_message.command);
  });
});

Toastbot.prototype.ensure_log_directory = method(function(self) {
  fs.mkdir(self.log_dir, '0777', function() {
    if(self.debug == true) {
      self.log('Ensured the log directory exists.');
    }
  });
});

Toastbot.prototype.get_date = method(function(self) {
  var today = new Date();
  return {
    today: today,
    year: today.getFullYear(),
    month: zero_pad(today.getMonth()),
    day: zero_pad(today.getDate()),
    hour: zero_pad(today.getHours()),
    minute: zero_pad(today.getMinutes()),
    second: zero_pad(today.getSeconds())
  };
});

Toastbot.prototype.log = method(function(self, message) {
  var date = self.get_date();
  
  var log_filename = '' + date.year + date.month + date.day + '.log';
  var log_filepath = path.join(self.log_dir, log_filename);
  
  fs.open(log_filepath, 'a', function(err, fd) {
    if(err) {
      console.log(err.toString());
      return;
    }
    
    var entry = '['+date.year+'-'+date.month+'-'+date.day+' '+date.hour+':'+date.minute+':'+date.second+'] ' + message+'\n';
    fs.write(fd, entry, function(err, written, buffer) {
      if(err) {
        console.log(err.toString());
      }
      fs.close(fd);
    });
  });
});

Toastbot.prototype.say = method(function(self, response) {
  if(response instanceof Array == false) {
    response = [response];
  }
  
  for(var roffset in response) {
    self.log(self.nick+': '+response[roffset]);
    self.client.say(self.channel, response[roffset]);
  }
});

Toastbot.prototype.clean_message = method(function(self, text) {
  var clean_text = text.trim();
  // Strip off crazy chars on actions.
  clean_text = clean_text.replace(/\u0001/g, '');
  return clean_text;
});

Toastbot.prototype.said_to_me = method(function(self, text) {
  for(var offset in self.variants) {
    if(text.indexOf(self.variants[offset]) == 0) {
      return ['direct', text.replace(self.variants[offset], '').trim()];
    }
    else if(text.indexOf(self.variants[offset]) > 0) {
      return ['indirect', text];
    }
    else {
      return ['nomention'];
    }
  }
});

Toastbot.prototype.is_direct_command = method(function(self, name, text) {
  var to_me = self.said_to_me(text);
  
  if(to_me[0] != 'direct') {
    return null;
  }
  
  // Use the modified text.
  text = to_me[1];
  
  if(text.toLowerCase() != name) {
    return null;
  }
  
  return text;
});

Toastbot.prototype.handle_message = method(function(self, nick, text) {
  var clean_text = self.clean_message(text);
  
  if(clean_text.substr(0, 6) == 'ACTION') {
    self.log('* '+nick+' '+clean_text.substr(5));
  }
  else {
    self.log(nick+": "+clean_text);
  }
 
  // Run the message through the handlers.
  for(var func_name in self) if(self[func_name].enabled && self[func_name].enabled()) {
    var response = self[func_name](nick, clean_text);
    
    if(response == null) {
      // Choosing not to handle this message.
      continue;
    }
    
    if(response == true) {
      // We're waiting on a callback. It'll handle the ``.say``.
      return;
    }
    
    // We got something back to say!
    self.say(response);
    return;

  }
});

Toastbot.prototype.handle_pm = method(function(self, nick, text) {
  var clean_text = self.clean_message(text);
  self.log("PM <- "+nick+": "+clean_text);
  var message = "Sorry, I don't respond to PMs yet.";
  self.client.say(nick, message);
  self.log("PM -> "+nick+": "+message);
});


// Exports.

exports.Toastbot = Toastbot;
exports.zero_pad = zero_pad;
exports.method = method;
exports.handler = handler;

// ==============================================
// Handlers!
// ==============================================


Toastbot.prototype.help = handler(method(function(self, nick, text) {
  var text = self.is_direct_command('help', text);
  
  if(! text) {
    return null;
  }
  
  var commands = [
    nick+": Valid commands -"
  ];
 

  for(var func_name in self) if(self[func_name].enabled && self[func_name].enabled()) {
    commands.push('   - '+func_name+' = '+(self[func_name].__doc__ || 'No documentation.'));
  } 
  
  return commands;
}));

Toastbot.prototype.help.__doc__ = "Provides a description of what I respond to.";

Toastbot.prototype.dance = handler(method(function(self, nick, text) {
  var text = self.is_direct_command('dance', text);
  
  if(! text) {
    return null;
  }
  
  var sweet_moves = [
    "_O_",
    "\\O_",
    "_O/",
    "\\O/",
  ];
  
  return sweet_moves;
}));
Toastbot.prototype.dance.__doc__ = "Get down and funky.";

Toastbot.prototype.wiki = handler(method(function(self, nick, text) {
  var to_me = self.said_to_me(text);
  
  if(to_me[0] != 'direct') {
    return null;
  }
  
  // Use the modified text.
  text = to_me[1];
  
  if(text.indexOf('wiki') != 0) {
    return null;
  }
  
  var search_terms = text.replace('wiki ', '');
  var options = {
    host: 'en.wikipedia.org',
    method: 'GET',
    path: '/w/index.php?search='+escape(search_terms),
    headers: {
      'User-Agent': 'Mozilla/4.0 (toastbot)'
    }
  };
  
  var req = http.get(options, function(res) {
    if(res.statusCode.toString() == '302') {
      self.say(nick+': '+res.headers['location']);
    }
  }).on('error', function(e) {
    self.log("Failed to load wiki entry for '"+search_terms+"': "+e.message);
  });
  
  return true;
}));
Toastbot.prototype.wiki.__doc__ = "Search Wikipedia for a topic.";

Toastbot.prototype.metar = handler(method(function(self, nick, text) {
  var to_me = self.said_to_me(text);
  
  if(to_me[0] != 'direct') {
    return null;
  }
  
  // Use the modified text.
  text = to_me[1];
  
  if(text.indexOf('metar') != 0) {
    return null;
  }
  
  var station = text.replace('metar ', '');
  var url = 'ftp://tgftp.nws.noaa.gov/data/observations/metar/stations/'+station.toUpperCase()+'.TXT';
  var curl = spawn('curl', [url]);
  var buffer = '';
  
  curl.stdout.on('data', function (data) {
    buffer += data;
  });
  
  curl.on('exit', function (code) {
    if(buffer.length > 0) {
      self.say(nick+': '+buffer.replace('\n', ' ').replace('\r', ''));
    }
    else {
      self.log("Failed to load metar entry for '"+station+"'.");
      self.say(nick+': Sorry, couldn\'t find that station.')
    }
  });
  
  return true;
}));
Toastbot.prototype.metar.__doc__ = "Fetch a NOAA METAR by station code.";

Toastbot.prototype.twitter = handler(method(function(self, nick, text) {
  var to_me = self.said_to_me(text);
  
  if(to_me[0] != 'direct') {
    return null;
  }
  
  // Use the modified text.
  text = to_me[1];
  
  if(text.indexOf('twitter') != 0) {
    return null;
  }
  
  var search_terms = text.replace('twitter ', '');
  var options = {
    host: 'search.twitter.com',
    method: 'GET',
    path: '/search.json?rpp=5&result_type=recent&q='+escape(search_terms),
    headers: {
      'User-Agent': 'Mozilla/4.0 (toastbot)'
    }
  };
  
  var req = http.get(options, function(res) {
    if(res.statusCode.toString() == '200') {
      var buffer = '';
      
      res.on('data', function(data) {
        buffer += data;
      });
      
      res.on('end', function() {
        var data = JSON.parse(buffer);
        var results = [
          nick+': Top 5 results'
        ];
        
        for(var offset in data['results']) {
          var tweet = data['results'][offset];
          results.push('   - @'+tweet.from_user+': '+tweet.text);
        }
        self.say(results);
      })
    }
    else {
      self.log("Failed to load Twitter search for '"+search_terms+"': "+res.statusCode);
      self.say(nick+': Sorry, Twitter isn\'t responding.');
    }
  }).on('error', function(e) {
    self.log("Failed to load wiki entry for '"+search_terms+"': "+e.message);
  });
  
  return true;
}));
Toastbot.prototype.twitter.__doc__ = "Search Twitter for a topic.";

File: toastbot.py
# -*- coding: utf-8 -*-
import irc

import codecs
import datetime
import os
import re
import socket
import subprocess
import urllib

import pyquery
import requests

try:
    import simplejson as json
except ImportError:
    import json


__author__ = 'Daniel Lindsley'
__version__ = (0, 4, 1)
__license__ = 'BSD'


class NotHandled(Exception):
    pass


class ToastBot(object):
    server = 'irc.freenode.net'
    port = 6667
    username = None
    realname = 'ToastBot'
    debug = False
    log_dir = os.path.join(os.path.dirname(__file__), 'logs')
    variant_endings = [
        ': ',
        ', ',
        '- ',
        ' - ',
    ]

    def __init__(self, nick, channel, **kwargs):
        self.nick = nick
        self.channel = channel
        self.client = None

        if 'server' in kwargs:
            self.server = kwargs['server']

        if 'port' in kwargs:
            self.port = kwargs['port']

        if 'username' in kwargs:
            self.username = kwargs['username']

        if 'realname' in kwargs:
            self.realname = kwargs['realname']

        if 'debug' in kwargs:
            self.debug = kwargs['debug']

        if 'log_dir' in kwargs:
            self.log_dir = kwargs['log_dir']

        self.variants = [self.nick + variant for variant in self.variant_endings]
        self.enabled_commands = [
            self.help,
            self.dance,
            self.woodies,
            self.wiki,
            self.metar,
            self.twitter,
            self.fatpita,
            self.corgibomb,
        ]

    def run(self):
        patterns = [
            (self.client.ping_re, self.client.handle_ping),
            (self.client.part_re, self.handle_part),
            (self.client.join_re, self.handle_join),
            (self.client.chanmsg_re, self.handle_channel_message),
            (self.client.privmsg_re, self.handle_private_message),
        ]
        self.client.logger.debug('entering receive loop')

        while 1:
            try:
                data = self.client._sock_file.readline()
            except socket.error:
                data = None

            if not data:
                self.client.logger.info('server closed connection')
                self.client.close()
                return True

            data = data.rstrip()

            for pattern, callback in patterns:
                match = pattern.match(data)
                if match:
                    callback(**match.groupdict())

    def setup(self):
        self.ensure_log_directory()
        self.client = irc.IRCConnection(self.server, self.port, self.nick, verbosity=2)
        self.client.connect()
        self.client.join(self.channel)
        self.run()

    def ensure_log_directory(self):
        if not os.path.exists(self.log_dir):
            os.makedirs(self.log_dir)

            if self.debug:
                self.log('Ensured the log directory exists.')

    def log(self, message):
        now = datetime.datetime.now()
        log_filename = "%4d%2d%2d.log" % (now.year, now.month, now.day)
        log_filepath = os.path.join(self.log_dir, log_filename)

        with codecs.open(log_filepath, 'a', encoding='utf-8') as log:
            log.write(u'[%4d-%02d-%02d %02d:%02d:%02d] %s\n' % (
                now.year,
                now.month,
                now.day,
                now.hour,
                now.minute,
                now.second,
                message
            ))

    def say(self, response):
        if not isinstance(response, (list, tuple)):
            response = [response]

        for resp in response:
            self.log(u"%s: %s" % (self.nick, resp))
            self.client.respond(resp.encode('utf-8'), channel=self.channel)

    def handle_join(self, nick, channel):
        self.log(u"%s joined %s." % (nick, channel))

    def handle_part(self, nick, channel):
        self.log(u"%s left %s." % (nick, channel))

    def clean_message(self, text):
        clean_text = text.strip()
        clean_text = clean_text.replace('\u0001', '')
        return clean_text

    def said_to_me(self, text):
        for variant in self.variants:
            if text.startswith(variant):
                return ['direct', text.replace(variant, '', 1)]
            elif variant in text:
                return ['indirect', text]

        return ['nomention', text]

    def is_direct_command(self, name, text):
        address, text = self.said_to_me(text)

        if address != 'direct':
            return None

        if text.lower() != name:
            return None

        return text

    def handle_channel_message(self, nick, channel, message):
        cleaned_text = self.clean_message(message)

        if cleaned_text.startswith('ACTION'):
            self.log(u"* %s %s" % (nick, cleaned_text.replace('ACTION', '', 1)))
        else:
            self.log(u"%s: %s" % (nick, cleaned_text))

        for command in self.enabled_commands:
            try:
                response = command(nick, cleaned_text)

                if response is True:
                    # It's doing it's own output.
                    return

                self.say(response)
            except NotHandled:
                # Nope, not that one. Try the next command.
                continue

    def handle_private_message(self, nick, text):
        cleaned_text = self.clean_message(text)
        self.log(u"PM <- %s: %s" % (nick, cleaned_text))
        message = "Sorry, I don't respond to PMs yet."
        self.log(u"PM -> %s: %s" % (nick, message))
        self.client.respond(message, nick=nick)

    # Available commands
    def help(self, nick, text):
        """Provides a description of what I respond to."""
        text = self.is_direct_command('help', text)

        if not text:
            raise NotHandled()

        commands = [
            u'%s: Valid commands - ' % nick,
        ]

        for command in self.enabled_commands:
            commands.append("  - %s = %s" % (command.__name__, command.__doc__ or 'No documentation.'))

        return commands

    def dance(self, nick, text):
        """Get down and funky."""
        text = self.is_direct_command('dance', text)

        if not text:
            raise NotHandled()

        sweet_moves = [
            "_O_",
            "\\O_",
            "_O/",
            "\\O/",
        ]
        return sweet_moves

    def woodies(self, nick, text):
        """Best quote on the internet."""
        if not 'woodies' in text:
            raise NotHandled()

        return 'U GUYZ R THE BEST AND GIVE ME A BILLION WOODIES A DAY! [https://code.djangoproject.com/ticket/7712#comment:2]'

    def wiki(self, nick, text):
        """Search Wikipedia for a topic."""
        address, text = self.said_to_me(text)

        if address != 'direct':
            raise NotHandled()

        if not text.startswith('wiki'):
            raise NotHandled()

        search_terms = text.replace('wiki ', '')
        resp = requests.get('http://en.wikipedia.org/w/index.php?search=%s' % urllib.quote_plus(search_terms), headers={'User-Agent': 'Mozilla/4.0 (toastbot)'})

        if resp.status_code in (404, 500):
            self.log("Failed to load wiki entry for '%s'." % search_terms)
            return True

        return u"%s: %s" % (nick, resp.url)

    def metar(self, nick, text):
        """Fetch a NOAA METAR by station code."""
        address, text = self.said_to_me(text)

        if address != 'direct':
            raise NotHandled()

        if not text.startswith('metar'):
            raise NotHandled()

        station = text.replace('metar ', '')
        url = "ftp://tgftp.nws.noaa.gov/data/observations/metar/stations/%s.TXT" % station.upper()
        proc = subprocess.Popen('curl %s' % url, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        stdout, stderr = proc.communicate()

        if proc.returncode != 0:
            self.log("Failed to load metar entry for '%s'." % station);
            return u"%s: Sorry, couldn't find that station." % nick

        return u"%s: %s" % (nick, stdout.replace('\n', ' ').replace('\r', ''))

    def twitter(self, nick, text):
        """Search Twitter for a topic."""
        address, text = self.said_to_me(text)

        if address != 'direct':
            raise NotHandled()

        if not text.startswith('twitter'):
            raise NotHandled()

        search_terms = text.replace('twitter ', '')
        resp = requests.get('http://search.twitter.com/search.json?rpp=5&result_type=recent&q=%s' % urllib.quote_plus(search_terms), headers={'User-Agent': 'Mozilla/4.0 (toastbot)'})

        if resp.status_code != 200:
            self.log("Failed to load wiki entry for '%s'." % search_terms)
            self.say(u"%s: Sorry, Twitter isn't responding." % nick)
            return True

        try:
            resp_data = json.loads(resp.content)
            results = [
                u'%s: Top 5 results - ' % nick,
            ]

            for tweet in resp_data.get('results', []):
                results.append(u"  - @%s: %s" % (tweet['from_user'], tweet['text']))

            return results
        except:
            self.log("FAIL WHALE for '%s'." % search_terms)
            self.say(u"%s: Twitter fail whale'd." % nick)
            return True

    def fatpita(self, nick, text):
        """Get a random fatpita image. For the lulz."""
        text = self.is_direct_command('fatpita', text)

        if not text:
            raise NotHandled()

        resp = requests.get('http://fatpita.net/', headers={'User-Agent': 'Mozilla/4.0 (toastbot)'})

        if resp.status_code in (404, 500):
            self.log("Failed to load random fatpita image.")
            return True

        return u"%s: %s" % (nick, resp.url)

    def corgibomb(self, nick, text):
        """CORGI BOMB!"""
        text = self.is_direct_command('corgibomb', text)

        if not text:
            raise NotHandled()

        resp = requests.get('http://www.tumblr.com/tagged/corgi', headers={'User-Agent': 'Mozilla/4.0 (toastbot)'})

        if resp.status_code in (404, 500):
            self.log("Failed to load corgibomb image.")
            return True

        doc = pyquery.PyQuery(resp.content)
        corgi_js = doc('.image_thumbnail:first').attr('onclick')

        # Because Tumblr LOL.
        tumblr_rage = re.search(r"this\.src=\'(?P<pic>.*?)\'", corgi_js)

        if tumblr_rage:
            corgi_pic = tumblr_rage.groupdict()['pic']
            return u"%s: %s" % (nick, corgi_pic)
        else:
            return u"%s: Sorry, Tumblr is being crappy. No pic for you." % nick
